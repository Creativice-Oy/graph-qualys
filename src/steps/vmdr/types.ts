/**
 * Maps a QWeb Host ID to a set of values it may be known by in
 * `Finding.targets`.
 *
 * Some values available on a Qualys `HostAsset` are not available on the
 * `DetectionHost`. The `HostAsset` values are collected in one step and made
 * available to a later step because the Host entities are only are not added to
 * the `jobState` except as mapped relationship `targetEntity` properties, and
 * are therefore not available for lookup later.
 */
export type HostAssetTargetsMap = Record<number, string[]>;

/**
 * Maps a Qualys vulnerability QID to the set of related `Finding._key` values.
 *
 * As Findings are created during the processing of `HostDetection` data, the
 * QID and `Finding._key` values are easily obtained, and therefore tracked for
 * a later step that will fetch vulnerabilities and create mapped relationships
 * between the Finding and Vulnerabilty.
 */
export type VulnerabilityFindingKeys = [number, Set<string>][];
