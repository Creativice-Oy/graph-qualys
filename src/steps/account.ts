import {
  createIntegrationEntity,
  IntegrationStep,
  IntegrationStepExecutionContext,
} from '@jupiterone/integration-sdk-core';

import { createQualysAPIClient } from '../provider';
import { QualysIntegrationConfig } from '../types';

export const STEP_FETCH_ACCOUNT = 'fetch-account';

export const ENTITY_TYPE_QUALYS_ACCOUNT = 'qualys_account';

export const DATA_ACCOUNT_ENTITY = 'DATA_ACCOUNT_ENTITY';

export async function fetchAccountDetails({
  logger,
  instance,
  jobState,
}: IntegrationStepExecutionContext<QualysIntegrationConfig>) {
  const apiClient = createQualysAPIClient(logger, instance.config);

  const data = await apiClient.fetchPortalInfo();
  const name = `Qualys - ${instance.name}`;

  const accountEntity = createIntegrationEntity({
    entityData: {
      source: {
        portal: data,
      },
      assign: {
        _type: ENTITY_TYPE_QUALYS_ACCOUNT,
        _class: 'Account',
        _key: `qualys-account:${instance.id}`,
        name,
        displayName: name,
      },
    },
  });

  await Promise.all([
    jobState.addEntity(accountEntity),
    jobState.setData(DATA_ACCOUNT_ENTITY, accountEntity),
  ]);
}

export const accountSteps: IntegrationStep<QualysIntegrationConfig>[] = [
  {
    id: STEP_FETCH_ACCOUNT,
    name: 'Fetch Account Details',
    entities: [
      {
        resourceName: 'Account',
        _type: ENTITY_TYPE_QUALYS_ACCOUNT,
        _class: 'Account',
        indexMetadata: {
          enabled: false,
        },
      },
    ],
    relationships: [],
    dependsOn: [],
    executionHandler: fetchAccountDetails,
  },
];
