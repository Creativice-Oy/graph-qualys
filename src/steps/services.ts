import {
  createDirectRelationship,
  createIntegrationEntity,
  Entity,
  IntegrationStep,
  IntegrationStepExecutionContext,
  JobState,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';

import { TYPE_QUALYS_ACCOUNT, TYPE_QUALYS_SERVICE } from '../converters';
import { createQualysAPIClient } from '../provider';
import { PortalInfo } from '../provider/client/types/portal';
import { QualysIntegrationConfig } from '../types';
import { DATA_ACCOUNT_ENTITY, STEP_FETCH_ACCOUNT } from './account';

export const STEP_FETCH_SERVICES = 'fetch-services';
export const DATA_WAS_SERVICE_ENTITY = 'DATA_WAS_SERVICE_ENTITY';
export const DATA_VMDR_SERVICE_ENTITY = 'DATA_VMDR_SERVICE_ENTITY';

export async function fetchServices({
  logger,
  instance,
  jobState,
}: IntegrationStepExecutionContext<QualysIntegrationConfig>) {
  const accountEntity = (await jobState.getData(DATA_ACCOUNT_ENTITY)) as Entity;

  const apiClient = createQualysAPIClient(logger, instance.config);
  const portalInfo = await apiClient.fetchPortalInfo();

  await Promise.all([
    createWebApplicationScannerService(jobState, accountEntity, portalInfo),
    createVulnerabilityManagementService(jobState, accountEntity, portalInfo),
  ]);
}

async function createWebApplicationScannerService(
  jobState: JobState,
  accountEntity: Entity,
  portalInfo?: PortalInfo,
): Promise<void> {
  const name = 'Qualys Web Application Scanner';
  const serviceEntity = await jobState.addEntity(
    createIntegrationEntity({
      entityData: {
        source: {},
        assign: {
          _type: TYPE_QUALYS_SERVICE,
          _class: 'Service',
          _key: `qualys-service:was`,
          displayName: name,
          name: name,
          category: ['software', 'other'],
          description:
            'Automated Web Application Security Assessment and Reporting',
          version: portalInfo?.['Portal-Version']?.['WAS-VERSION'],
          createdOn: undefined,
          updatedOn: undefined,
        },
      },
    }),
  );

  await jobState.addRelationship(
    createDirectRelationship({
      _class: RelationshipClass.HAS,
      from: accountEntity,
      to: serviceEntity,
    }),
  );

  await jobState.setData(DATA_WAS_SERVICE_ENTITY, serviceEntity);
}

async function createVulnerabilityManagementService(
  jobState: JobState,
  accountEntity: Entity,
  portalInfo?: PortalInfo,
): Promise<void> {
  const name = 'Qualys Vulnerability Manager';
  const serviceEntity = await jobState.addEntity(
    createIntegrationEntity({
      entityData: {
        source: {},
        assign: {
          _type: TYPE_QUALYS_SERVICE,
          _class: 'Service',
          _key: `qualys-service:vmdr`,
          displayName: name,
          name: name,
          category: ['software', 'other'],
          description:
            'Detect, prioritize and remediate vulnerabilities, and monitor using dashboards.',
          version: portalInfo?.['Portal-Version']?.['VM-VERSION'],
          createdOn: undefined,
          updatedOn: undefined,
        },
      },
    }),
  );

  await jobState.addRelationship(
    createDirectRelationship({
      _class: RelationshipClass.HAS,
      from: accountEntity,
      to: serviceEntity,
    }),
  );

  await jobState.setData(DATA_VMDR_SERVICE_ENTITY, serviceEntity);
}

export const serviceSteps: IntegrationStep<QualysIntegrationConfig>[] = [
  {
    id: STEP_FETCH_SERVICES,
    name: 'Fetch Services',
    entities: [
      {
        resourceName: 'Service',
        _type: TYPE_QUALYS_SERVICE,
        _class: 'Service',
      },
    ],
    relationships: [
      {
        _type: 'qualys_account_has_service',
        _class: RelationshipClass.HAS,
        sourceType: TYPE_QUALYS_ACCOUNT,
        targetType: TYPE_QUALYS_SERVICE,
      },
    ],
    dependsOn: [STEP_FETCH_ACCOUNT],
    executionHandler: fetchServices,
  },
];
