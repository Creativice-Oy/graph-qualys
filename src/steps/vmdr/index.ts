import {
  createDirectRelationship,
  Entity,
  getRawData,
  IntegrationStep,
  IntegrationStepExecutionContext,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';

import { createQualysAPIClient } from '../../provider';
import { QualysIntegrationConfig } from '../../types';
import {
  Steps,
  VmdrEntities,
  VmdrMappedRelationships,
  VmdrRelationships,
} from './constants';
import { createHostEntity } from './converters';
import { DATA_ACCOUNT_ENTITY, STEP_FETCH_ACCOUNT } from '../account';
import { Host } from '../../provider/client/types/vmpc/listHosts';

export async function fetchHosts({
  logger,
  instance,
  jobState,
}: IntegrationStepExecutionContext<QualysIntegrationConfig>) {
  const apiClient = createQualysAPIClient(logger, instance.config);
  const accountEntity = (await jobState.getData(DATA_ACCOUNT_ENTITY)) as Entity;

  await apiClient.iterateHosts(async (host) => {
    const hostEntity = await jobState.addEntity(createHostEntity(host));

    await jobState.addRelationship(
      createDirectRelationship({
        _class: RelationshipClass.HAS,
        from: accountEntity,
        to: hostEntity,
      }),
    );
  });
}

export async function mapHostToEc2GcpRelationship({
  jobState,
}: IntegrationStepExecutionContext<QualysIntegrationConfig>) {
  await jobState.iterateEntities(
    { _type: VmdrEntities.HOST._type },
    async (hostEntity) => {
      const host = getRawData<Host>(hostEntity);

      if (host) {
        console.log(host);
      }

      return Promise.resolve();
    },
  );
}

export const hostDetectionSteps: IntegrationStep<QualysIntegrationConfig>[] = [
  {
    id: Steps.HOSTS,
    name: 'Fetch Hosts',
    entities: [VmdrEntities.HOST],
    relationships: [VmdrRelationships.ACCOUNT_HAS_HOST],
    dependsOn: [STEP_FETCH_ACCOUNT],
    executionHandler: fetchHosts,
  },
  {
    id: Steps.BUILD_HOST_MAPPED_RELATIONSHIP,
    name: 'Build Host and GCP/EC2 Relationship',
    entities: [],
    relationships: [],
    mappedRelationships: [
      VmdrMappedRelationships.HOST_IS_EC2,
      VmdrMappedRelationships.HOST_IS_GCP,
    ],
    dependsOn: [Steps.HOSTS],
    executionHandler: mapHostToEc2GcpRelationship,
  },
];
