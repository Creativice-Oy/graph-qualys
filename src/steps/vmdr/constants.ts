import {
  generateRelationshipType,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';

import { ENTITY_TYPE_SERVICE_VMDR } from '../services';

// Step IDs
export const STEP_FETCH_SCANNED_HOST_IDS = 'fetch-scanned-host-ids';
export const STEP_FETCH_SCANNED_HOST_DETAILS = 'fetch-scanned-host-details';
export const STEP_FETCH_SCANNED_HOST_FINDINGS = 'fetch-scanned-host-detections';

// Inter-step data storage keys
export const DATA_SCANNED_HOST_IDS = 'DATA_SCANNED_HOST_IDS';
export const DATA_HOST_TARGETS = 'DATA_HOST_TARGETS';
export const DATA_VULNERABILITY_FINDING_KEYS =
  'DATA_VULNERABILITY_FINDING_KEYS';

export const ENTITY_TYPE_HOST_FINDING = 'qualys_host_finding';

export const ENTITY_TYPE_DISCOVERED_HOST = 'discovered_host';
export const ENTITY_TYPE_EC2_HOST = 'aws_instance';

export const RELATIONSHIP_TYPE_SERVICE_FINDING = generateRelationshipType(
  RelationshipClass.IDENTIFIED,
  ENTITY_TYPE_SERVICE_VMDR,
  ENTITY_TYPE_HOST_FINDING,
);

export const MAPPED_RELATIONSHIP_TYPE_VDMR_DISCOVERED_HOST = generateRelationshipType(
  RelationshipClass.SCANS,
  ENTITY_TYPE_SERVICE_VMDR,
  ENTITY_TYPE_DISCOVERED_HOST,
);
export const MAPPED_RELATIONSHIP_TYPE_VDMR_EC2_HOST = generateRelationshipType(
  RelationshipClass.SCANS,
  ENTITY_TYPE_SERVICE_VMDR,
  ENTITY_TYPE_EC2_HOST,
);

export const VmdrEntities = {
  HOST_FINDING: {
    _type: ENTITY_TYPE_HOST_FINDING,
    _class: 'Finding',
    resourceName: 'Detection',
  },
};

export const VmdrRelationships = {
  SERVICE_FINDING: {
    _type: RELATIONSHIP_TYPE_SERVICE_FINDING,
    _class: RelationshipClass.IDENTIFIED,
    sourceType: ENTITY_TYPE_SERVICE_VMDR,
    targetType: ENTITY_TYPE_HOST_FINDING,
  },
  SERVICE_DISCOVERED_HOST: {
    _type: MAPPED_RELATIONSHIP_TYPE_VDMR_DISCOVERED_HOST,
    _class: RelationshipClass.SCANS,
    sourceType: ENTITY_TYPE_SERVICE_VMDR,
    targetType: ENTITY_TYPE_DISCOVERED_HOST,
  },
  SERVICE_EC2_HOST: {
    _type: MAPPED_RELATIONSHIP_TYPE_VDMR_EC2_HOST,
    _class: RelationshipClass.SCANS,
    sourceType: ENTITY_TYPE_SERVICE_VMDR,
    targetType: ENTITY_TYPE_EC2_HOST,
  },
};
