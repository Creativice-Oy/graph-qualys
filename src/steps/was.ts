import {
  createIntegrationEntity,
  Entity,
  IntegrationStep,
  IntegrationStepExecutionContext,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';

import { buildWebAppKey, TYPE_QUALYS_WEB_APP } from '../converters';
import { createQualysAPIClient } from '../provider';
import { QualysIntegrationConfig } from '../types';
import { DATA_WAS_SERVICE_ENTITY, STEP_FETCH_SERVICES } from './services';

export async function fetchWebApps({
  logger,
  instance,
  jobState,
}: IntegrationStepExecutionContext<QualysIntegrationConfig>) {
  const apiClient = createQualysAPIClient(logger, instance.config);

  const serviceEntity = (await jobState.getData(
    DATA_WAS_SERVICE_ENTITY,
  )) as Entity;

  // TODO Process vulnerabilities, relate to serviceEntity
  await apiClient.iterateWebApps(async (data) => {
    await jobState.addEntity(
      createIntegrationEntity({
        entityData: {
          source: data as any,
          assign: {
            _type: TYPE_QUALYS_WEB_APP,
            _key: buildWebAppKey({
              webAppId: data.id!,
            }),
            _class: 'Application',
            id: String(data.id),
            name: data.name!,
            displayName: data.name!,
            createdOn: parseTimePropertyValue(data.createdDate),
            updatedOn: parseTimePropertyValue(data.updatedDate),
          },
        },
      }),
    );
  });
}

export const webApplicationSteps: IntegrationStep<QualysIntegrationConfig>[] = [
  {
    id: 'fetch-webapps',
    name: 'Fetch Web Apps',
    entities: [
      {
        _type: TYPE_QUALYS_WEB_APP,
        _class: 'Application',
        resourceName: 'Web App',
      },
    ],
    relationships: [],
    dependsOn: [STEP_FETCH_SERVICES],
    executionHandler: fetchWebApps,
  },
];
