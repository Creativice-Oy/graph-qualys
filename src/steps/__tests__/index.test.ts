import {
  createMockStepExecutionContext,
  Recording,
} from '@jupiterone/integration-sdk-testing';

import { config } from '../../../test/config';
import { setupQualysRecording } from '../../../test/recording';
import { QualysIntegrationConfig } from '../../types';
import { fetchAccountDetails } from '../account';
import { fetchServices } from '../services';
import { fetchHostDetections, fetchHostIds } from '../vmdr';
import { fetchWebApps } from '../was';

jest.setTimeout(10000 * 2);

let recording: Recording;

afterEach(async () => {
  await recording.stop();
});

test('steps', async () => {
  recording = setupQualysRecording({
    directory: __dirname,
    name: 'steps',
  });

  const context = createMockStepExecutionContext<QualysIntegrationConfig>({
    instanceConfig: config,
  });

  await fetchAccountDetails(context);
  await fetchServices(context);
  await fetchWebApps(context);
  await fetchHostIds(context);
  await fetchHostDetections(context);

  // Review snapshot, failure is a regression
  expect({
    numCollectedEntities: context.jobState.collectedEntities.length,
    numCollectedRelationships: context.jobState.collectedRelationships.length,
    collectedEntities: context.jobState.collectedEntities,
    collectedRelationships: context.jobState.collectedRelationships,
    encounteredTypes: context.jobState.encounteredTypes,
  }).toMatchSnapshot();

  expect(
    context.jobState.collectedEntities.filter((e) =>
      e._class.includes('Account'),
    ),
  ).toMatchGraphObjectSchema({
    _class: ['Account'],
    schema: {
      additionalProperties: true,
      properties: {
        _type: { const: 'qualys_account' },
        name: { type: 'string' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['name'],
    },
  });

  expect(
    context.jobState.collectedEntities.filter((e) =>
      e._class.includes('Service'),
    ),
  ).toMatchGraphObjectSchema({
    _class: ['Service'],
    schema: {
      additionalProperties: true,
      properties: {
        _type: { const: 'qualys_service' },
        name: { type: 'string' },
        version: { type: 'string' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['name', 'version'],
    },
  });
});
