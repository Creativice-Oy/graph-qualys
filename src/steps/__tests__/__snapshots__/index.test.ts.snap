// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchScannedHostDetails: fetchScannedHostDetails 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "qualys-account:local-integration-instance",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "portal": Object {
              "Portal-Version": Object {
                "AV2-VERSION": "0.1.0",
                "CA-VERSION": "2.23.0.0",
                "CERTVIEW-VERSION": "2.7.0.0-67",
                "CLOUDVIEW-VERSION": "1.11.0.0-1",
                "CM-VERSION": "1.30.0.0",
                "CS-VERSION": "1.6.8.2-23",
                "FIM-VERSION": "2.3.1.0",
                "IOC-VERSION": "3.0.0.0-5",
                "ITAM-VERSION": "1.2.1.1-1",
                "MDS-VERSION": "2.15.7.0",
                "PM-VERSION": "1.4.3.0-289",
                "PS-VERSION": "1.2.1.0-4",
                "PortalApplication-VERSION": "3.1.2.0-2 OFFICIAL #597 (2020-09-15T15:54:59Z)",
                "QGS-VERSION": "1.2.0.0-6",
                "QUESTIONNAIRE-VERSION": "2.23.1.0",
                "QUESTIONNAIRE__V2-VERSION": "1.10.1.0",
                "THREAT__PROTECT-VERSION": "1.2.0.0",
                "UD-VERSION": "0.0.6",
                "VM-VERSION": "1.0.3",
                "WAF-VERSION": "2.12.4.0",
                "WAS-VERSION": "6.13.1.0",
              },
              "QWeb-Version": Object {
                "SCANNER-VERSION": "12.1.67-1",
                "VULNSIGS-VERSION": "2.5.13-2",
                "WEB-VERSION": "10.3.0.3-2",
              },
            },
          },
        },
      ],
      "_type": "qualys_account",
      "createdOn": undefined,
      "displayName": "Qualys - Local Integration",
      "name": "Qualys - Local Integration",
    },
    Object {
      "_class": Array [
        "Service",
      ],
      "_key": "qualys-service:was",
      "_rawData": Array [],
      "_type": "qualys_web_app_scanner",
      "category": Array [
        "software",
        "other",
      ],
      "createdOn": undefined,
      "description": "Automated Web Application Security Assessment and Reporting",
      "displayName": "Qualys Web Application Scanner",
      "function": "DAST",
      "name": "Qualys Web Application Scanner",
      "updatedOn": undefined,
      "version": "6.13.1.0",
    },
    Object {
      "_class": Array [
        "Service",
      ],
      "_key": "qualys-service:vmdr",
      "_rawData": Array [],
      "_type": "qualys_vulnerability_manager",
      "category": Array [
        "software",
        "other",
      ],
      "createdOn": undefined,
      "description": "Detect, prioritize and remediate vulnerabilities, and monitor using dashboards.",
      "displayName": "Qualys Vulnerability Manager",
      "function": "vulnerability-management",
      "name": "Qualys Vulnerability Manager",
      "updatedOn": undefined,
      "version": "1.0.3",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "qualys-account:local-integration-instance",
      "_key": "qualys-account:local-integration-instance|has|qualys-service:was",
      "_toEntityKey": "qualys-service:was",
      "_type": "qualys_account_has_web_app_scanner",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "qualys-account:local-integration-instance",
      "_key": "qualys-account:local-integration-instance|has|qualys-service:vmdr",
      "_toEntityKey": "qualys-service:vmdr",
      "_type": "qualys_account_has_vulnerability_manager",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": Array [
    "qualys_account",
    "qualys_web_app_scanner",
    "qualys_vulnerability_manager",
    "qualys_account_has_web_app_scanner",
    "qualys_account_has_vulnerability_manager",
  ],
  "numCollectedEntities": 3,
  "numCollectedRelationships": 2,
}
`;

exports[`fetchScannedHostFindings: fetchScannedHostFindings 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "qualys-account:local-integration-instance",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "portal": Object {
              "Portal-Version": Object {
                "AV2-VERSION": "0.1.0",
                "CA-VERSION": "2.23.0.0",
                "CERTVIEW-VERSION": "2.7.0.0-67",
                "CLOUDVIEW-VERSION": "1.11.0.0-1",
                "CM-VERSION": "1.30.0.0",
                "CS-VERSION": "1.6.8.2-23",
                "FIM-VERSION": "2.3.1.0",
                "IOC-VERSION": "3.0.0.0-5",
                "ITAM-VERSION": "1.2.1.1-1",
                "MDS-VERSION": "2.15.7.0",
                "PM-VERSION": "1.4.3.0-289",
                "PS-VERSION": "1.2.1.0-4",
                "PortalApplication-VERSION": "3.1.2.0-2 OFFICIAL #597 (2020-09-15T15:54:59Z)",
                "QGS-VERSION": "1.2.0.0-6",
                "QUESTIONNAIRE-VERSION": "2.23.1.0",
                "QUESTIONNAIRE__V2-VERSION": "1.10.1.0",
                "THREAT__PROTECT-VERSION": "1.2.0.0",
                "UD-VERSION": "0.0.6",
                "VM-VERSION": "1.0.3",
                "WAF-VERSION": "2.12.4.0",
                "WAS-VERSION": "6.13.1.0",
              },
              "QWeb-Version": Object {
                "SCANNER-VERSION": "12.1.67-1",
                "VULNSIGS-VERSION": "2.5.13-2",
                "WEB-VERSION": "10.3.0.3-2",
              },
            },
          },
        },
      ],
      "_type": "qualys_account",
      "createdOn": undefined,
      "displayName": "Qualys - Local Integration",
      "name": "Qualys - Local Integration",
    },
    Object {
      "_class": Array [
        "Service",
      ],
      "_key": "qualys-service:was",
      "_rawData": Array [],
      "_type": "qualys_web_app_scanner",
      "category": Array [
        "software",
        "other",
      ],
      "createdOn": undefined,
      "description": "Automated Web Application Security Assessment and Reporting",
      "displayName": "Qualys Web Application Scanner",
      "function": "DAST",
      "name": "Qualys Web Application Scanner",
      "updatedOn": undefined,
      "version": "6.13.1.0",
    },
    Object {
      "_class": Array [
        "Service",
      ],
      "_key": "qualys-service:vmdr",
      "_rawData": Array [],
      "_type": "qualys_vulnerability_manager",
      "category": Array [
        "software",
        "other",
      ],
      "createdOn": undefined,
      "description": "Detect, prioritize and remediate vulnerabilities, and monitor using dashboards.",
      "displayName": "Qualys Vulnerability Manager",
      "function": "vulnerability-management",
      "name": "Qualys Vulnerability Manager",
      "updatedOn": undefined,
      "version": "1.0.3",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "qualys-account:local-integration-instance",
      "_key": "qualys-account:local-integration-instance|has|qualys-service:was",
      "_toEntityKey": "qualys-service:was",
      "_type": "qualys_account_has_web_app_scanner",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "qualys-account:local-integration-instance",
      "_key": "qualys-account:local-integration-instance|has|qualys-service:vmdr",
      "_toEntityKey": "qualys-service:vmdr",
      "_type": "qualys_account_has_vulnerability_manager",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": Array [
    "qualys_account",
    "qualys_web_app_scanner",
    "qualys_vulnerability_manager",
    "qualys_account_has_web_app_scanner",
    "qualys_account_has_vulnerability_manager",
  ],
  "numCollectedEntities": 3,
  "numCollectedRelationships": 2,
}
`;

exports[`steps 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "qualys-account:local-integration-instance",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "portal": Object {
              "Portal-Version": Object {
                "AV2-VERSION": "0.1.0",
                "CA-VERSION": "2.23.0.0",
                "CERTVIEW-VERSION": "2.7.0.0-67",
                "CLOUDVIEW-VERSION": "1.11.0.0-1",
                "CM-VERSION": "1.30.0.0",
                "CS-VERSION": "1.6.8.2-23",
                "FIM-VERSION": "2.3.1.0",
                "IOC-VERSION": "3.0.0.0-5",
                "ITAM-VERSION": "1.2.1.1-1",
                "MDS-VERSION": "2.15.7.0",
                "PM-VERSION": "1.4.3.0-289",
                "PS-VERSION": "1.2.0.0-RELEASE",
                "PortalApplication-VERSION": "3.1.2.0-2 OFFICIAL #597 (2020-09-15T15:54:59Z)",
                "QGS-VERSION": "1.2.0.0-6",
                "QUESTIONNAIRE-VERSION": "2.23.1.0",
                "QUESTIONNAIRE__V2-VERSION": "1.10.1.0",
                "THREAT__PROTECT-VERSION": "1.2.0.0",
                "UD-VERSION": "0.0.6",
                "VM-VERSION": "1.0.3",
                "WAF-VERSION": "2.12.4.0",
                "WAS-VERSION": "6.13.1.0",
              },
              "QWeb-Version": Object {
                "SCANNER-VERSION": "12.1.67-1",
                "VULNSIGS-VERSION": "2.5.4-2",
                "WEB-VERSION": "10.3.0.3-2",
              },
            },
          },
        },
      ],
      "_type": "qualys_account",
      "createdOn": undefined,
      "displayName": "Qualys - Local Integration",
      "name": "Qualys - Local Integration",
    },
    Object {
      "_class": Array [
        "Service",
      ],
      "_key": "qualys-service:was",
      "_rawData": Array [],
      "_type": "qualys_web_app_scanner",
      "category": Array [
        "software",
        "other",
      ],
      "createdOn": undefined,
      "description": "Automated Web Application Security Assessment and Reporting",
      "displayName": "Qualys Web Application Scanner",
      "function": "DAST",
      "name": "Qualys Web Application Scanner",
      "updatedOn": undefined,
      "version": "6.13.1.0",
    },
    Object {
      "_class": Array [
        "Service",
      ],
      "_key": "qualys-service:vmdr",
      "_rawData": Array [],
      "_type": "qualys_vulnerability_manager",
      "category": Array [
        "software",
        "other",
      ],
      "createdOn": undefined,
      "description": "Detect, prioritize and remediate vulnerabilities, and monitor using dashboards.",
      "displayName": "Qualys Vulnerability Manager",
      "function": "vulnerability-management",
      "name": "Qualys Vulnerability Manager",
      "updatedOn": undefined,
      "version": "1.0.3",
    },
    Object {
      "_class": Array [
        "Finding",
      ],
      "_key": "c9d59cd1-8d7b-4cd4-8cdc-19cf09d5fd5c",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "uploadStatus": "SKIPPED",
            "uploadStatusReason": "Raw data for detection entities currently disabled",
          },
        },
      ],
      "_type": "qualys_web_app_finding",
      "category": "app-scan",
      "createdOn": 1599696500000,
      "displayName": "Missing header: Referrer-Policy",
      "id": "3158304",
      "lastTestedOn": 1599696500000,
      "name": "Missing header: Referrer-Policy",
      "numericSeverity": 4,
      "open": true,
      "qid": 150208,
      "severity": "low",
      "status": undefined,
      "targets": Array [
        "Ask JupiterOne",
      ],
      "type": "INFORMATION_GATHERED",
      "updatedOn": 1599696500000,
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "qualys-account:local-integration-instance",
      "_key": "qualys-account:local-integration-instance|has|qualys-service:was",
      "_toEntityKey": "qualys-service:was",
      "_type": "qualys_account_has_web_app_scanner",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "qualys-account:local-integration-instance",
      "_key": "qualys-account:local-integration-instance|has|qualys-service:vmdr",
      "_toEntityKey": "qualys-service:vmdr",
      "_type": "qualys_account_has_vulnerability_manager",
      "displayName": "HAS",
    },
    Object {
      "_class": "SCANS",
      "_key": "qualys-service:was|scans|FORWARD:_type=web_app:name=Ask JupiterOne",
      "_mapping": Object {
        "relationshipDirection": "FORWARD",
        "sourceEntityKey": "qualys-service:was",
        "targetEntity": Object {
          "_class": "Application",
          "_type": "web_app",
          "displayName": "Ask JupiterOne",
          "name": "Ask JupiterOne",
        },
        "targetFilterKeys": Array [
          Array [
            "_type",
            "name",
          ],
        ],
      },
      "_type": "qualys_web_app_scanner_scans_web_app",
      "displayName": "SCANS",
    },
    Object {
      "_class": "IDENTIFIED",
      "_fromEntityKey": "qualys-service:was",
      "_key": "qualys-service:was|identified|c9d59cd1-8d7b-4cd4-8cdc-19cf09d5fd5c",
      "_toEntityKey": "c9d59cd1-8d7b-4cd4-8cdc-19cf09d5fd5c",
      "_type": "qualys_web_app_scanner_identified_finding",
      "displayName": "IDENTIFIED",
    },
    Object {
      "_class": "SCANS",
      "_key": "qualys-service:vmdr|scans|qualys-host:107800671",
      "_mapping": Object {
        "relationshipDirection": "FORWARD",
        "sourceEntityKey": "qualys-service:vmdr",
        "targetEntity": Object {
          "_class": "Host",
          "_key": "arn:aws:ec2:us-east-1:191097756250:instance/i-02751173c463c34e8",
          "_type": "aws_instance",
          "displayName": "i-02751173c463c34e8",
          "fqdn": "ec2-3-236-88-110.compute-1.amazonaws.com",
          "hostname": "ec2-3-236-88-110.compute-1.amazonaws.com",
          "id": "i-02751173c463c34e8",
          "ipAddress": "172.31.66.126",
          "lastScannedOn": undefined,
          "os": "Linux",
          "platform": "linux",
          "privateIpAddress": "172.31.66.126",
          "publicIpAddress": undefined,
          "qualysAssetId": 84047683,
          "qualysHostId": 107800671,
          "scannedBy": "qualys",
        },
        "targetFilterKeys": Array [
          Array [
            "_type",
            "_key",
          ],
        ],
      },
      "_type": "qualys_vulnerability_manager_scans_aws_instance",
      "displayName": "SCANS",
    },
    Object {
      "_class": "IS",
      "_key": "c9d59cd1-8d7b-4cd4-8cdc-19cf09d5fd5c|is|vuln-qid:150208",
      "_mapping": Object {
        "relationshipDirection": "FORWARD",
        "sourceEntityKey": "c9d59cd1-8d7b-4cd4-8cdc-19cf09d5fd5c",
        "targetEntity": Object {
          "_class": "Vulnerability",
          "_key": "vuln-qid:150208",
          "_type": "qualys_vuln",
          "displayName": "Missing header: Referrer-Policy",
          "id": "150208",
          "name": "Missing header: Referrer-Policy",
          "severityLevel": 2,
          "webLink": "https://qualysguard.qg3.apps.qualys.com/fo/common/vuln_info.php?id=150208",
        },
        "targetFilterKeys": Array [
          Array [
            "_type",
            "_key",
          ],
        ],
      },
      "_type": "qualys_web_app_finding_is_vuln",
      "displayName": "IS",
    },
  ],
  "encounteredTypes": Array [
    "qualys_account",
    "qualys_web_app_scanner",
    "qualys_vulnerability_manager",
    "qualys_account_has_web_app_scanner",
    "qualys_account_has_vulnerability_manager",
    "qualys_web_app_scanner_scans_web_app",
    "qualys_web_app_finding",
    "qualys_web_app_scanner_identified_finding",
    "qualys_vulnerability_manager_scans_aws_instance",
    "qualys_web_app_finding_is_vuln",
  ],
  "numCollectedEntities": 4,
  "numCollectedRelationships": 6,
}
`;
